Use parameters inside rules when calling commands

giig1967g opened the pull request:
Hi,
here is a PR for using parameters when calling commands from inside the rules.
This will save a lot of text thus reducing processing time during text file rule parsing.
**In order to call the evaluate the formula, a prefix of '=' must be used. See examples.**

The scope is to be able to use formulas instead of an integer when calling a command inside the rules.

Examples:

1)
taskvalueset,1,=%eventvalue%+65,=[Relay1#r1]+1

2)
Assuming you have a PCF8574 with ports from 65 to 72, you can write a rule:

```
on turnOnRelay do
  pcfgpio,=%eventvalue%+64,0
endon

```
And to turn on port 67, the rule can be called like this:

`event,turnOnRelay=3`

giig1967g committed:
Use parameters inside rules when calling commands

Hi,
here is a PR for using parameters when calling commands from inside the rules.

The scope is to be able to use formulas instead of an integer when calling a command inside the rules.

Example:
taskvalueset,1,%eventvalue%+65,[Relay1#r1]+1

Usage could be:

Assuming you have a PCF8574 with ports from 65 to 72.

You can write a rule:

on turnOnRelay do
  pcfgpio,%eventvalue%+64,0
endon

And to turn on port65, the rule can be called like this:

event,turnOn=1

TD-er commented:
I think this is indeed a great (and simple) addition.
Only thing is, can you add a bit more descriptive info to the log regarding the return value of the Calculate function?

giig1967g commented:
Sure. Tomorrow will add it.

giig1967g committed:
few changes

optimized the call to CALCULATE only when needed and description of error codes.

giig1967g committed:
fixed ESP32 compatibility issue

giig1967g commented:
I have added a check in order to call the CALCULATE function only if needed.

TD-er commented on src/Misc.ino:
Please add F() macro.
Otherwise it will use a lot of memory.

giig1967g committed:
fixed calls to F-macro

giig1967g commented:
done

giig1967g commented:
@TD-er: a quick advice from you:

in order to make sure that the CALCULATE function is called only if needed (to avoid to slow down rules) , I came up with the following idea:

Call the CALCULATE function ONLY if the formula has a prefix of '='.
So the syntax would be:
`taskvalueset,=%eventvalue%+1,1,=[dummy1#value1]+10`

If there is no '=' the Calculate will not be called and previous logic will be used:
`taskvalueset,1,1,1`

This is the safest mode, as the user is aware that he is using the functionality, and this need to be documented in the wiki, otherwise no one will know it.
Maybe is less intuitive, but for sure there is no risk to call the function when it is not needed.

What do you think?

TD-er commented:
The calculate function could also check the string for that.
I think it is a proper prefix to mark calculations.

giig1967g commented:
The CALCULATE function is called from many different places, not only from rules.
Other cases are the formula fields inside the devices page.
So for backward compatibility I wouldn't make the '=' mandatory inside the calculate() function, but only inside the new function calculateParam() that calls calculate(), and inside calculate() if first char is '=' then it should just be discarted.

Will send the code changed in a short time.

giig1967g committed:
final version with '=' sign as a prefix to identify the formula in rules
